File: binary trees/recursive/ConvertSortedArraytoBinarySearchTree.py
class Solution(object):
    def sortedArrayToBST(self, nums):
        
        if len(nums) == 0:
            return None
        elif len(nums) == 1:
            return TreeNode(nums[0], None, None)
        elif len(nums) == 2:
            return TreeNode(nums[0], None, TreeNode(nums[1], None, None))
        
        mid = int(len(nums) / 2)
        left = self.sortedArrayToBST(nums[:mid])
        right = self.sortedArrayToBST(nums[mid+1:])
        
        return TreeNode(nums[mid], left, right)
            
            
# divide arrays / subarrays in binary search process:
# find middle index (rounded down bc of division flooring)
# left = root returned by recursive call on left side
# right = root returned by recursive call on right side

File: binary trees/recursive/ConvertSortedArraytoBinarySearchTree.java
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        if(nums == null || nums.length == 0) return null; 
        if(nums.length == 1) return new TreeNode(nums[0]);
        if(nums.length == 2) return new TreeNode(nums[0], null, new TreeNode(nums[1]));
        
        int mid = nums.length / 2;
        TreeNode left = sortedArrayToBST(Arrays.copyOfRange(nums, 0, mid));
        TreeNode right = sortedArrayToBST(Arrays.copyOfRange(nums, mid+1, nums.length));
        
        return new TreeNode(nums[mid], left, right);
    }
}
